cirq-core/cirq/transformers/dynamical_decoupling.py

```diff
@@ -98,112 +111,92 @@ def _is_single_qubit_gate_moment(moment: 'cirq.Moment') -> bool:
-    remaining_pulled_through = pulled_through
-    for affected_q, combined_op_in_pauli in pulled_through.items():
-        moment_id = mutable_circuit.prev_moment_operating_on([affected_q], clifford_piece_range[1])
-        if moment_id is not None:
-            op = mutable_circuit.operation_at(affected_q, moment_id)
-            # Try to merge op into an existing single-qubit gate operation.
-            if op is not None and _is_single_qubit_operation(op):
-                updated_gate_mat = cirq.unitary(combined_op_in_pauli) @ cirq.unitary(op)
-                updated_gate: Optional['cirq.Gate'] = (
-                    single_qubit_decompositions.single_qubit_matrix_to_phxz(updated_gate_mat)
-                )
-                if updated_gate is None:
-                    # updated_gate is close to Identity.
-                    updated_gate = cirq.I
-                batch_replaces.append((moment_id, op, updated_gate.on(affected_q)))
-                remaining_pulled_through *= combined_op_in_pauli.on(affected_q)
-                continue
-            # Insert into the first empty moment for the qubit if such moment exists.
-            while moment_id < clifford_piece_range[1]:
-                if affected_q not in mutable_circuit.moments[
-                    moment_id
-                ].qubits and _is_insertable_moment(
-                    mutable_circuit.moments[moment_id], single_qubit_gate_moments_only
-                ):
-                    insert_intos.append((moment_id, combined_op_in_pauli.on(affected_q)))
-                    remaining_pulled_through *= combined_op_in_pauli.on(affected_q)
-                    break
-                moment_id += 1
-    mutable_circuit.batch_insert_into(insert_intos)
-    mutable_circuit.batch_replace(batch_replaces)
-    return remaining_pulled_through
+    clifford_ops_in_moment: list['cirq.Operation'] = [
+        op for op in moment.operations if _is_clifford_op(op)
+    ]
+    return input_pauli_ops.after(clifford_ops_in_moment)
+
+
+def _get_stop_qubits(moment: 'cirq.Moment') -> set['cirq.Qid']:
+    stop_pulling_through_qubits: set['cirq.Qid'] = set()
+    for op in moment:
+        if (not _is_clifford_op(op) and not _is_single_qubit_operation(op)) or not cirq.has_unitary(
+            op
+        ):  # multi-qubit clifford op or non-mergable op.
+            stop_pulling_through_qubits.update(op.qubits)
+    return stop_pulling_through_qubits
+
+
+def _need_merge_pulled_through(op_at_q: 'cirq.Operation', is_at_last_busy_moment: bool) -> bool:
+    """With a pulling through puali gate before op_at_q, need to merge with the
+    pauli in the conditions below."""
+    # The op must be mergable and single-qubit
+    if not (cirq.has_unitary(op_at_q) and _is_single_qubit_operation(op_at_q)):
+        return False
+    # Either non-Clifford or at the last busy moment
+    return (not _is_clifford_op(op_at_q)) or is_at_last_busy_moment
```

Comment:
swap for performance
```suggestion
    return is_at_last_busy_moment or not _is_clifford_op(op_at_q)
```